Implement send and receive text and file

bryanchriswhite opened the pull request:
This PR implements a dart, object-oriented API that wraps the cgo API that is exported by `wormhole-william` in accordance with https://github.com/LeastAuthority/dart_wormhole_william/issues/7.

Depends on https://github.com/LeastAuthority/wormhole-william/pull/26

Additionally, #3 and #4 are implemented in this PR.

---

## Code Review Checklist (to be filled out by reviewer)

- [x] Description accurately reflects what changes are being made.
- [x] Description explains why the changes are being made (or references an issue containing one).
- [ ] The PR appropriately sized.
- [ ] New code has enough tests. 
- [ ] New code has enough documentation to answer "how do I use it?" and "what does it do?".
- [x] Existing documentation is up-to-date if impacted.

bryanchriswhite committed:
wip -- BDDing send and recv text

bryanchriswhite committed:
update wormhole-william

bryanchriswhite committed:
wip

ewanas committed:
Test statically linking the cgo code to the shared library

ewanas committed:
Fix cmake compilation to use the go runtime.
Add android NDK specific handling for building the go library

ewanas committed:
Fix cgo invocation to work in Windows

ewanas committed:
TEMP dart sendfile

ewanas committed:
Update gitignore

ewanas committed:
Revert dependency updates

ewanas committed:
Revert cmake changes for linux

bryanchriswhite committed:
not what we want to do!

bryanchriswhite committed:
updated wormhole-william

bryanchriswhite committed:
wip

ewanas committed:
TEMP

ewanas committed:
callbacks wip

ewanas committed:
Compile the contents of the entire ./c/ tree instead of just the client

ewanas committed:
Try calling dart port sender directly without wormhole_william interaction

ewanas committed:
Use FLUTTER_HOME When resolving dart c files

bryanchriswhite committed:
wip

bryanchriswhite committed:
dart cleanup

bryanchriswhite committed:
bugfixing in cmake and c src

bryanchriswhite committed:
include lib/src/CMakeLists.txt in android/CMakeLists.txt

bryanchriswhite committed:
fix my mistakes

bryanchriswhite committed:
more cleanup

bryanchriswhite committed:
more cleanup

ewanas committed:
Try to send to a dart port from C

bryanchriswhite committed:
add missing dart vm init call

bryanchriswhite committed:
cleanup

ewanas committed:
Add send text implementation

ewanas committed:
fixup! Add send text implementation

bryanchriswhite committed:
wip

bryanchriswhite committed:
wip

bryanchriswhite committed:
wip

bryanchriswhite committed:
cleanup

bryanchriswhite committed:
uncomment and update sendFile

bryanchriswhite committed:
uncomment error handling

bryanchriswhite committed:
cleanup

bryanchriswhite committed:
wip

bryanchriswhite committed:
add exceptions

ewanas committed:
Separate handling for different callbacks

bryanchriswhite committed:
fix dumb and broken things

ewanas committed:
Add debug print functionality

bryanchriswhite committed:
cleanup and update submodule

bryanchriswhite committed:
make sendFile take a File object

bryanchriswhite committed:
add client receive file members

ewanas committed:
Comment out old client usage from test. Possibly remove this test later

ewanas committed:
Add options to native client constructor

ewanas committed:
Fix client after rebasing on client with tranit relay CLI option

bryanchriswhite committed:
update wormhole-william

bryanchriswhite committed:
wip debugging

bryanchriswhite committed:
fix typos

bryanchriswhite committed:
update wormhole-william

ewanas committed:
Fix Android/Linux shared library builds

ewanas committed:
Set default URLs to use the public servers instead of local environments

ewanas commented on android/CMakeLists.txt:
```diff
@@ -23,9 +27,8 @@ else()
   set(GOARCH "please-fail :/")
```
Definitely should use a fail command like the one used for `FLUTTER_HOME`

ewanas replied:
Added explicit failure here

ewanas commented on android/CMakeLists.txt:
```diff
@@ -23,9 +27,8 @@ else()
+    ${ANDROID_TOOLCHAIN_ROOT}/bin/${CMAKE_C_COMPILER_TARGET}-clang)
 
 # Fix the clang triplet used as parameter for -target flag. compiler with -none-
 # are not in the ndk
```
Figure out if we could clean these text replacements on `CC_FOR_TARGET`

ewanas replied:
Tried removing either one and the issue remains; these replacements seem needed for now

ewanas commented on android/CMakeLists.txt:
```diff
@@ -35,6 +38,12 @@ string(REPLACE "-none" "" CC_FOR_TARGET ${CC_FOR_TARGET})
 string(REPLACE "armv7-" "armv7a-" CC_FOR_TARGET ${CC_FOR_TARGET})
 
-add_library(${PLUGIN_NAME} SHARED "src/main/c/dummy.c")
+add_library(${PLUGIN_LIB} SHARED "src/main/c/dummy.c")
```
See if we can avoid having a dummy C file

ewanas replied:
Seems like we're stuck with this hack for now. The issue is that linux(and maybe other platforms) have code within that library, but for Android we didn't have such a need.

ewanas commented on android/CMakeLists.txt:
```diff
@@ -35,6 +38,12 @@ string(REPLACE "-none" "" CC_FOR_TARGET ${CC_FOR_TARGET})
+include(${DART_WORMHOLE_WILLIAM_C_DIR}/../../CMakeLists.txt)
+include(${DART_WORMHOLE_WILLIAM_C_DIR}/CMakeLists.txt)
+
+target_link_libraries(${PLUGIN_LIB} PRIVATE bindings)
```
Cleanup here instead of two commands

ewanas commented on android/CMakeLists.txt:
```diff
@@ -35,6 +38,12 @@ string(REPLACE "-none" "" CC_FOR_TARGET ${CC_FOR_TARGET})
+target_link_libraries(${PLUGIN_LIB} PRIVATE wormhole_william)
 
-include(../CMakeLists.txt)
+set(dart_wormhole_william_bundled_libraries "" $<TARGET_FILE:wormhole_william>
```
Double check about that `""`

ewanas replied:
Removed

bryanchriswhite commented on lib/client/client.dart:
```diff
@@ -0,0 +1,159 @@
+          final exception =
+              Exception('Failed to send text. Error code: $errCode');
+          done.completeError(exception);
+          throw exception;
```
Does this `throw` make sense?

ewanas replied:
Removed

bryanchriswhite commented on lib/client/client.dart:
```diff
@@ -0,0 +1,159 @@
+      // TODO: Create exception implementation(s).
+      calloc.free(codeOut);
+      final exception = Exception('Failed to send text. Error code: $errCode');
+      done.completeError(exception);
```
Don't think this makes sense.

ewanas replied:
Removed

bryanchriswhite commented on lib/client/client.dart:
```diff
@@ -0,0 +1,159 @@
+    _config = config;
+  }
+
+  Future<SendResult> sendText(String msg) {
```
Make this actually async.

bryanchriswhite commented on lib/client/client.dart:
```diff
@@ -0,0 +1,159 @@
+          final exception =
+              Exception('Failed while sending text. Error code: $response');
+          done.completeError(exception);
+          throw exception;
```
```suggestion
```

bryanchriswhite commented on lib/client/client.dart:
```diff
@@ -0,0 +1,159 @@
+    return Future.value(result);
+  }
+
+  Future<String> recvText(String code) {
```
Make actually async

bryanchriswhite commented on lib/client/client.dart:
```diff
@@ -0,0 +1,159 @@
+  Future<SendResult> sendFile(File file) async {
+    final fileName = path.basename(file.path);
+    final length = await file.length();
+    print("File length was: $length");
```
```suggestion
```

bryanchriswhite commented on lib/client/client.dart:
```diff
@@ -0,0 +1,159 @@
+    print("File length was: $length");
+    final done = Completer<void>();
+
+    Pointer<Pointer<Utf8>> codeOut = calloc();
```
We should look into this, what is `calloc` doing exactly?

ewanas replied:
To reiterate our joint digging for this one, `calloc` is an `Allocator` and allocators are callable due to an extension function defined on `Allocator` that takes an optional count(default is 1).

bryanchriswhite commented on lib/client/client.dart:
```diff
@@ -0,0 +1,159 @@
+
+    final rxPort = ReceivePort()
+      ..listen((dynamic errCode) {
+        print("Got ${errCode}");
```
```suggestion
```

bryanchriswhite commented on lib/client/client.dart:
```diff
@@ -0,0 +1,159 @@
+        print("Got ${errCode}");
+        if (errCode is int) {
+          // TODO: Create exception implementation(s).
+          throw Exception(
```
This should call `completeError` on the completer and not throw an exception.

bryanchriswhite commented on lib/client/exceptions.dart:
```diff
@@ -0,0 +1,3 @@
+enum ErrorCodes {
```
I don't think we use this. Clean up.

bryanchriswhite commented on lib/client/native_client.dart:
```diff
@@ -0,0 +1,178 @@
+    Pointer<Utf8> transitRelayUrl,
+    Int32 passPhraseComponentLength);
+typedef NewClient = int Function(
+    Pointer<Utf8> appId,
```
Can / should these be `String`?

bryanchriswhite commented on lib/client/native_client.dart:
```diff
@@ -0,0 +1,178 @@
+typedef ClientRecvFile = int Function(
+    int goClientId, Pointer<Utf8> code, int callbackPortId);
+
+const DEFAULT_APP_ID = "lothar.com/wormhole/text-or-file-xfer";
```
Allow override via environment variables or config file at build-time, rely on go for defaults.

bryanchriswhite commented on lib/client/native_client.dart:
```diff
@@ -0,0 +1,178 @@
+  late final DynamicLibrary _asyncCallbackLib;
+  late final _goClientId;
+
+  String _appId = DEFAULT_APP_ID;
+  String _rendezvousUrl = DEFAULT_RENDEZVOUS_URL;
+  String _transitRelayUrl = DEFAULT_TRANSIT_RELAY_URL;
+  int _passPhraseComponentLength = DEFAULT_PASSPHRASE_COMPONENT_LENGTH;
```
Simplify. We can use a new / default config in the constructor.

ewanas replied:
Did a bit of refactoring to clean this up. `NativeClient` now either uses the provided `Config` or creates a `Config` with no explicit parameters, thereby using the defaults.

bryanchriswhite commented on lib/client/native_context.dart:
```diff
@@ -0,0 +1,6 @@
+import 'dart:ffi';
+
+class NativeContext extends Struct {
```
Double-check if we're using this.

bryanchriswhite commented on lib/src/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+                  $ENV{FLUTTER_HOME}/bin/cache/dart-sdk/include/dart_api_dl.c)
+target_include_directories(
+  bindings PRIVATE $ENV{FLUTTER_HOME}/bin/cache/dart-sdk/include)
+target_include_directories(bindings PRIVATE ${flutter_INCLUDE_DIRECTORIES})
```
Double-check if needed or redundant.

ewanas replied:
Removed. The preceding include directory is sufficient.

bryanchriswhite commented on lib/src/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+set_target_properties(
+  bindings
+  PROPERTIES PUBLIC_HEADER bindings.h
+             PUBLIC_HEADER include/dart_version.h
```
Do these need to be `PUBLIC_HEADER`s?

ewanas replied:
Removed. They were superfluous

bryanchriswhite commented on lib/src/async_callback.c:
```diff
@@ -0,0 +1,136 @@
+
+#include "libwormhole_william.h"
+// TODO define this in the build
+#define DEBUG
```
Move to CMakeLists

ewanas replied:
Implementation added to the bindings CMakeLists.txt file.

ewanas committed:
Remove debug print calls

Co-authored-by: Bryan White <bryanchriswhite@gmail.com>

ewanas committed:
Cleanup to build scripts. Fix for linux build issue with missing wormhole headers

ewanas committed:
Cleanup default configuration handling

ewanas committed:
Remove blocking calls from client functions.
Avoid throwing exceptions after completeError.
Remove redundant complete calls after completeError is called

ewanas committed:
Fix some flutter analyze issues

ewanas committed:
Return the result_t struct to dart

ewanas committed:
Avoid copying the result. Dart will free when it is done with it

ewanas committed:
Remove unused files and dead code

ewanas committed:
Show logs from the C code only for debug builds
